packages/react-native/ReactCommon/react/nativemodule/core/platform/ios/ReactCommon/RCTTurboModule.mm

```diff
@@ -400,12 +481,77 @@ id convertJSIValueToObjCObject(jsi::Runtime &runtime, const jsi::Value &value, s
-      throw convertNSExceptionToJSError(runtime, exception);
-    } @finally {
-      [retainedObjectsForInvocation removeAllObjects];
+    std::optional<jsi::JSError> handledCatch = std::nullopt;
+    try {
+      @try {
+        [inv invokeWithTarget:strongModule];
+       } @catch (NSException *exception) {
+         handledCatch = handleCause(runtime,
+                     isSync,
+                     @{
+                       @"name": exception.name,
+                       @"message": exception.reason,
+                       @"stackSymbols": exception.callStackSymbols,
+                       @"stackReturnAddresses": exception.callStackReturnAddresses,
+                     },
+                     retainedObjectsForInvocation,
+                     optionalInternalRejectBlock);
+       } @catch (NSError *error) {
+         handledCatch = handleCause(runtime,
+                     isSync,
+                     @{
+                       @"userInfo": error.userInfo,
+                       @"domain": error.domain,
+                     },
+                     retainedObjectsForInvocation,
+                     optionalInternalRejectBlock);
+       } @catch (NSString *errorMessage) {
+         handledCatch = handleCause(runtime,
+                     isSync,
+                     @{
+                       @"message": errorMessage,
+                     },
+                     retainedObjectsForInvocation,
+                     optionalInternalRejectBlock);
+       } @catch (id e) {
+         handledCatch = handleCause(runtime,
+                     isSync,
+                     @{
+                       @"message": @"Unknown Objective-C Object thrown.",
+                     },
+                     retainedObjectsForInvocation,
+                     optionalInternalRejectBlock);
+       } @finally {
+         [retainedObjectsForInvocation removeAllObjects];
+       }
+    } catch (const std::exception &exception) {
+      handledCatch = handleCause(runtime,
+                  isSync,
+                  @{
+                    @"message": [NSString stringWithUTF8String:exception.what()],
+                  },
+                  retainedObjectsForInvocation,
+                  optionalInternalRejectBlock);
+    } catch (const std::string &errorMessage) {
+      handledCatch = handleCause(runtime,
+                  isSync,
+                  @{
+                    @"message": [NSString stringWithUTF8String:errorMessage.c_str()],
+                  },
+                  retainedObjectsForInvocation,
+                  optionalInternalRejectBlock);
+    } catch (...) {
+      handledCatch = handleCause(runtime,
+                  isSync,
+                  @{
+                    @"message": @"Unknown C++ exception thrown.",
+                  },
+                  retainedObjectsForInvocation,
+                  optionalInternalRejectBlock);
+    }
+
+    if (handledCatch.has_value()) {
+      throw handledCatch.value();
     }
```

Comment:
I think you want some logic here that catches if it's safe, but re-throws otherwise.

```
    NSDictionary *caughtException = nil;
    BOOL shouldCatchException = isSync || optionalInternalRejectBlock;
    try {
      @try {
        [inv invokeWithTarget:strongModule];
       } @catch (NSException *exception) {
         caughtException = maybeCatchException(shouldCatchException, @{
                       @"name": exception.name,
                       @"message": exception.reason,
                       @"stackSymbols": exception.callStackSymbols,
                       @"stackReturnAddresses": exception.callStackReturnAddresses,
                     });
       } ...
    } catch { ... }

    if (caughtException) {
      if (isSync) {
        throw convertNSDictionaryToJSError(runtime, caughtException);
      } else {
        optionalInternalRejectBlock(caughtException);
      }
    }
```